nio/api.py

```diff
@@ -234,11 +234,42 @@ def encrypted_mxc_to_plumb(mxc, key, hash, iv, homeserver=None):
         return plumb_url
 
     @staticmethod
-    def _build_path(path, query_parameters=None, api_path=MATRIX_API_PATH):
-        # type: (str, dict, str) -> str
-        path = ("{api}/{path}").format(api=api_path, path=path)
+    def _build_path(
+        path: List[str],
+        query_parameters: Dict = None,
+        base_path: str = MATRIX_API_PATH
+    ) -> str:
+        """Builds a percent-encoded path from a list of strings.
 
-        path = quote(path)
+        For example, turns ["hello", "wo/rld"] into "/hello/wo%2Frld".
+        All special characters are percent encoded,
+        including the forward slash (/).
+
+        Args:
+            path (List[str]): the list of path elements.
+            query_parameters (Dict, optional): [description]. Defaults to None.
+            base_path (str, optional): A base path to be prepended to path. Defaults to MATRIX_API_PATH.
+
+        Returns:
+            str: [description]
+        """
+        quoted_path: str = None
+
+        if isinstance(path, str):
+            quoted_path = quote(path, safe="")
+        elif isinstance(path, List):
+            quoted_path='/'.join([quote(str(part), safe="") for part in path])
+        else:
+            quoted_path = path
+
+        path = "{base}/{path}".format(
+            base=base_path,
+            path=quoted_path
+        )
+
+        # remove trailing /
+        if path[-1] == "/":
```

Comment:
I don't think that check is necessary, `rstrip` will be a noop if the last char isn't a `/`, no?